source/python/idetect/model.py

```diff
@@ -65,35 +65,76 @@ class Article(Base):
     reports = relationship('Report')
 
-    def update_status(self, new_status):
+    def __str__(self):
+        return "<Article {} {} {}>".format(self.id, self.url_id, self.url)
+
+    def get_updated_version(self):
+        """Return the most recent version of this article by updated timestamp"""
+        return Article.get_latest_version(object_session(self), self.url_id)
+
+    @classmethod
+    def get_latest_version(cls, session, url_id=None, url=None, lock=False):
+        """
+        Return the Article with the given url_id or url with the most recent updated timestamp.
+        If lock is True, uses FOR UPDATE locking, so that other calls to get_latest_version will
+        block until the transaction is committed or rolled back.
         """
-        Atomically Update the status of this Article from to new_status.
-        If something changed the status of this article since it was loaded, raise.
+        query = session.query(cls)
+        if url_id is not None:
+            query = query.filter(Article.url_id == url_id)
+        if url is not None:
+            query = query.filter(Article.url == url)
+        # if neither url_id nor url is provided, this returns the most recent article overall
+        if lock:
+            # This prevents this being run again until the transaction is committed or rolled back
+            query = query.with_for_update()
+        return query.order_by(desc(Article.updated)).limit(1).first()
+
+    def create_new_version(self, new_status):
+        """
+        Try to create a new version of this article with the new status.
+        If this is not the most recent version for this
+        url_id, this will raise NotLatestException.
         """
         session = object_session(self)
-        if not session:
-            raise RuntimeError("Object has not been persisted in a session.")
-
-        expected_status = self.status
-        result = session.query(Article).filter(Article.id == self.id, Article.status == self.status) \
-            .update(
-            {
-                Article.status: new_status
-            })
-        if result != 1:
-            try:
-                updated = session.query(Article).filter(
-                    Article.id == self.id).one()
-                raise UnexpectedArticleStatusException(
-                    self, expected_status, updated.status)
-            except NoResultFound:
-                raise UnexpectedArticleStatusException(
-                    self, expected_status, None)
+        try:
+            if not session:
+                raise RuntimeError("Object has not been persisted in a session.")
+
+            # lock to prevent simultaneous updates
+            latest = Article.get_latest_version(session, self.url_id, lock=True)
+            if latest.id != self.id:
+                raise NotLatestException(self, latest)
+
+            make_transient(self)
+            self.id = None
```

Comment:
Cool. We'll need to figure out all the different scenarios that should/shouldn't copy the mappings.
I think we had also said that in some scenarios we would copy the extracted reports too.